src/protocols/bdx/AsyncTransferFacilitator.cpp

```diff
@@ -0,0 +1,201 @@
+            // or an error occured, we need to call the CleanUp virtual method implemented by the subclass to clean up and
+            // delete both the sublcass and base class objects.
+            if (outEvent.EventType == TransferSession::OutputEventType::kAckEOFReceived ||
+                outEvent.EventType == TransferSession::OutputEventType::kInternalError ||
+                outEvent.EventType == TransferSession::OutputEventType::kTransferTimeout ||
+                outEvent.EventType == TransferSession::OutputEventType::kStatusReceived)
+            {
+                CleanUp();
+                return err;
+            }
+        }
+        mTransfer.GetNextAction(outEvent);
+    };
+    return err;
+}
+
+void AsyncTransferFacilitator::OnExchangeClosing(Messaging::ExchangeContext * ec)
+{
+    ChipLogDetail(BDX, "OnExchangeClosing, ec: " ChipLogFormatExchange, ChipLogValueExchange(ec));
+    mExchange.Release();
+}
+
+void AsyncTransferFacilitator::OnResponseTimeout(Messaging::ExchangeContext * ec)
+{
+    ChipLogDetail(BDX, "OnResponseTimeout, ec: " ChipLogFormatExchange, ChipLogValueExchange(ec));
+    CleanUp();
+}
+
+CHIP_ERROR AsyncResponder::PrepareForTransfer(Messaging::ExchangeContext * exchangeCtx, TransferRole role,
+                                              BitFlags<TransferControlFlags> xferControlOpts, uint16_t maxBlockSize,
+                                              System::Clock::Timeout timeout)
+{
+    VerifyOrReturnError(exchangeCtx != nullptr, CHIP_ERROR_INCORRECT_STATE);
+    VerifyOrReturnError(!mExchange, CHIP_ERROR_INCORRECT_STATE);
+
+    mTimeout = timeout;
+
+    ReturnErrorOnFailure(mTransfer.WaitForTransfer(role, xferControlOpts, maxBlockSize, mTimeout));
+
+    mExchange.Grab(exchangeCtx);
+
+    return CHIP_NO_ERROR;
+}
+
+bdx::StatusCode AsyncResponder::GetBdxStatusCodeFromChipError(CHIP_ERROR err)
+{
+    if (err == CHIP_ERROR_INCORRECT_STATE)
+    {
+        return bdx::StatusCode::kUnexpectedMessage;
+    }
+    if (err == CHIP_ERROR_INVALID_ARGUMENT)
+    {
+        return bdx::StatusCode::kBadMessageContents;
+    }
+    return bdx::StatusCode::kUnknown;
+}
+
+void AsyncResponder::NotifyEventHandled(CHIP_ERROR eventHandlingResult)
+{
+    // We can only come here after we have handled a BDX message (like ReceiveInit/BlockQuery) asynchronously
+    // in the subclass. The base class needs to notify us when the handling of the event is done and the event
+    // handling result - success or failure.
+    ChipLogDetail(BDX, "NotifyEventHandled : error %" CHIP_ERROR_FORMAT, eventHandlingResult.Format());
+
+    // If there was an error, we abort the transfer.
+    if (eventHandlingResult != CHIP_NO_ERROR)
+    {
+        CleanUp();
+    }
+    else
+    {
+
+        // We need to get the next output event from the state machine which is a response to the
+        // BDX message handled by the subclass (ReceiveAccept/Block).
```

Comment:
I don't get it.  We didn't put anything into the state machine here.  Why would we expect to be getting anything out of it?  Is the expectation that our subclass will mess with `mTransfer` to cause it to have new events?  Because that's not documented in our API anywhere....  And neither is the fact that our API consumers don't need to worry about kMsgToSend events.

Reply:
This is to handle the asynchronous cases as described in the comment above.